certbot/tests/error_handler_test.py

```diff
@@ -30,25 +50,20 @@ def test_context_manager(self):
         self.init_func.assert_called_once_with(*self.init_args,
                                                **self.init_kwargs)
 
-    @mock.patch('certbot.error_handler.os')
-    @mock.patch('certbot.error_handler.signal')
-    def test_signal_handler(self, mock_signal, mock_os):
-        # pylint: disable=protected-access
-        mock_signal.getsignal.return_value = signal.SIG_DFL
-        self.handler.set_signal_handlers()
-        signal_handler = self.handler._signal_handler
-        for signum in self.signals:
-            mock_signal.signal.assert_any_call(signum, signal_handler)
-
-        signum = self.signals[0]
-        signal_handler(signum, None)
+    def test_context_manager_with_signal(self):
+        with signal_receiver(self.signals) as signals_received:
+            with self.handler:
+                should_be_42 = 42
+                send_signal(signal.SIGTERM)
+                should_be_42 *= 10
+
+        # check exectuion stoped when the signal was sent
+        assert 42 == should_be_42
+        # assert signals were caught
+        assert [signal.SIGTERM] == signals_received
+        # assert the error handling function was just called once
         self.init_func.assert_called_once_with(*self.init_args,
                                                **self.init_kwargs)
-        mock_os.kill.assert_called_once_with(mock_os.getpid(), signum)
-
-        self.handler.reset_signal_handlers()
-        for signum in self.signals:
-            mock_signal.signal.assert_any_call(signum, signal.SIG_DFL)
```

Comment:
Removing this, there is now no test that we restore signals correctly. There are a number of ways to do this, but I think in general, we should basically duplicate what the `ErrorHandler` does in a test. We should use `signal.getsignal` to record all signal handlers in use before using the `ErrorHandler` and use `self.assertEqual` to check that the signal was restored properly after using the `ErrorHandler`.